specs/phase1/beacon-chain.md

```diff
@@ -1057,51 +538,124 @@ def process_epoch(state: BeaconState) -> None:
     process_reveal_deadlines(state)
     process_challenge_deadlines(state)
     process_slashings(state)
+    process_pending_headers(state)
+    charge_confirmed_header_fees(state)
+    reset_pending_headers(state)
     process_final_updates(state)  # phase 0 final updates
     process_phase_1_final_updates(state)
 ```
 
+#### Pending headers
+
+```python
+
+def process_pending_headers(state: BeaconState):
+    for slot in range(SLOTS_PER_EPOCH):
+        for shard in range(get_active_shard_count(state)):
+            # Pending headers for this (slot, shard) combo
+            candidates = [
+                c for c in state.previous_epoch_pending_shard_headers if
+                (c.slot, c.shard) == (slot, shard)
+            ]
+            # The entire committee (and its balance)
+            full_committee = get_beacon_committee(state, slot, shard)
+            full_committee_balance = get_total_balance(state, full_committee)
+            if True not in [c.confirmed for c in candidates]:
+                # The set of voters who voted for each header
+                # (and their total balances)
+                voting_sets = [
+                    [v for i, v in enumerate(full_committee) if c.votes[i]]
+                    for c in candidates
+                ]
+                voting_balances = [
+                    get_total_balance(state, voters)
+                    for voters in voting_sets
+                ]
+                # Get the index with the most total balance voting for them.
+                # NOTE: if two choices get exactly the same voting balance,
+                # the candidate earlier in the list wins
+                if max(voting_balances) > 0:
+                    winning_index = voting_balances.index(max(voting_balances))
+                else:
+                    # If no votes, zero wins
+                    winning_index = [c.root for c in candidates].index(Root())
```

Comment:
Should we also cover the case when no proposals were made for a slot, hence no such index will be found?